Attach pid

tcourdy opened the pull request:
Closes #30

tcourdy committed:
attach_pid example is working in the debug library.  Also moved the functions to get the process information from either a linux or windows machine to the debug library

tcourdy committed:
plugged attach_pid example code into the actual server endpoints that the client will call to attach via pid

rpjohnst commented on server/Cargo.toml:
```diff
@@ -15,6 +15,7 @@ serde_json = "0.8"
 debug = { path = "debug" }
 winapi = { git = "https://github.com/team-worm/winapi-rs" }
 kernel32-sys = { git = "https://github.com/team-worm/winapi-rs" }
+advapi32-sys = { git = "https://github.com/team-worm/winapi-rs" }
```
`server` shouldn't depend on `kernel32-sys` or `advapi-sys` (or ideally `winapi`, but `debug` still exposes some of that).

rpjohnst commented on server/debug/examples/attach_pid.rs:
```diff
@@ -0,0 +1,285 @@
+    let pid = env::args().nth(1).unwrap().parse::<u32>().unwrap();
+
+    unsafe {
+        let h_process = kernel32::OpenProcess(
```
This (and all the other `advapi32`/`kernel32` calls in this file) should be a part of the `debug` crate rather than the example (and thus the `server` crate). That way we can keep all the platform-specific stuff in one place, to be reimplemented for Linux later on.

rpjohnst commented on server/debug/examples/attach_pid.rs:
```diff
@@ -0,0 +1,285 @@
+
+        // set debug privileges
+
+        let mut token_handle: winapi::winnt::HANDLE = mem::zeroed();
```
This is done in several places, but `winapi::winnt::HANDLE` should be shortened to either `winapi::HANDLE` or `std::os::windows::io::RawHandle` (the second, when in examples like here or the `server` crate).

rpjohnst commented on server/debug/examples/attach_pid.rs:
```diff
@@ -0,0 +1,285 @@
+
+// I don't think these will be needed since debugActiveProcess will call the appropriate debug events
+// for all modules and threads that the process had started or loaded before we broke in.
+// These may be useful helper functions later though.
```
Ah, that is great. I did not realize `DebugActiveProcess` did that- should probably just delete these in that case.

rpjohnst commented on server/debug/src/process.rs:
```diff
@@ -54,6 +55,10 @@ impl Child {
         Ok(())
     }
+
+    pub fn new(handle: winapi::winnt::HANDLE) -> io::Result<Child> {
```
This should be called `attach`, take a `RawHandle` rather than a `winapi::HANDLE`, and should be where all the setup happens, up to and including `DebugActiveProcess`. That way we can reimplement it for other platforms later without changing code outside of the `debug` crate.

rpjohnst commented on server/debug/src/process.rs:
```diff
@@ -265,3 +270,87 @@ fn ensure_no_nuls<S: AsRef<OsStr>>(s: S) -> io::Result<S> {
+}
+
+/// helper function to get list of running windows processes on machine
+pub fn read_win_proc() -> Vec<Proc> {
```
Could you mark these with `#[cfg(windows)]` and `#[cfg(unix)]` and give them the same name?

It would also be good for them to return an `Iterator` rather than a `Vec`.

rpjohnst commented on server/src/child.rs:
```diff
@@ -88,17 +108,93 @@ struct DebugState {
     last_breakpoint: Option<usize>,
 }
 
+fn attach_pid(pid: u32, tx: SyncSender<DebugMessage>, rx: Receiver<ServerMessage>) -> io::Result<()> {
```
Maybe just call this `attach`?

In any case, this looks like the same code from the example you wrote that needs to be moved into the `debug` crate. I.e. this should just call `Child::attach` instead of making winapi calls itself.

rpjohnst reviewed (changes requested):
Functionality looks good; needs some reorganization.

tcourdy committed:
Reorganized code based off of suggestions from pull request

rpjohnst committed:
Attach-by-PID cleanup

- clean up the example to do what the server does
- simplify some win32 API usage
- use OsStr{,ing} and Path when interfacing with platform strings
- make `list_running_processes` a method of `Process`
- use an iterator to enumerate processes
- combine common code from `Thread::launch` and `Thread::attach`

rpjohnst committed:
Merge branch 'master' into attach_pid
